SingleConnectionFactory - reconnection problem (AMQ Broker)

I have a problem with reconnection when `SingleConnectionFactory` (`CachingConnectionFactory`) is used.
I think it is a similar (or same?) problem reported in #23058 issue.

From all sources, which I have (logs, heap dump ...), it seems that problem is caused due **missing Exception Listener
on connection (`ActiveMQConnection`), which is stored on `SingleConnectionFactory`**.

I have following environment:
- Apache ActiveMQ Broker 5.16.2
- `failover:(tcp://localhost:61616)?maxReconnectAttempts=5`
- `SingleConnectionFactory` (`CachingConnectionFactory`)
 - `setReconnectOnException(true)`

###### Logs
- it is visible that "resetting the underlying JMS Connection" on `CachingConnectionFactory` is invoked
- but suddenly these logs are no longer present and only logs from `DefaultMessageListenerContainer` are present
  -  I know that it is not recommended to use DMLC with `CachingConnectionFactory`, but please stay with me for a moment (because I think the problem is in `SingleConnectionFactory`)

```
2022-09-06T05:16:06.024+00:00 {ActiveMQ Connection Executor: unconnected} [CachingConnectionFactory] [INFO] Encountered a JMSException - resetting the underlying JMS Connection
...
2022-09-06T05:18:37.210+00:00 {ActiveMQ Connection Executor: unconnected} [CachingConnectionFactory] [INFO] Encountered a JMSException - resetting the underlying JMS Connection
...
2022-09-06T05:21:08.254+00:00 {ActiveMQ Connection Executor: unconnected} [CachingConnectionFactory] [INFO] Encountered a JMSException - resetting the underlying JMS Connection
...
2022-09-06T05:23:39.319+00:00 {ActiveMQ Connection Executor: unconnected} [CachingConnectionFactory] [INFO] Encountered a JMSException - resetting the underlying JMS Connection
...
2022-09-06T05:26:15.367+00:00 {ActiveMQ Connection Executor: unconnected} [CachingConnectionFactory] [INFO] Encountered a JMSException - resetting the underlying JMS Connection
...
2022-09-06T05:26:25.498+00:00 {taskExecutor-151} [DefaultMessageListenerContainer] [ERROR] Could not refresh JMS Connection for destination '...' - retrying using FixedBackOff{interval=5000, currentAttempts=1, maxAttempts=unlimited}. Cause: The JMS connection has failed: ...cluster.local
2022-09-06T05:26:30.499+00:00 {taskExecutor-151} [DefaultMessageListenerContainer] [ERROR] Could not refresh JMS Connection for destination '...' - retrying using FixedBackOff{interval=5000, currentAttempts=2, maxAttempts=unlimited}. Cause: The JMS 
...
2022-09-06T05:46:30.559+00:00 {taskExecutor-142} [DefaultMessageListenerContainer] [ERROR] Could not refresh JMS Connection for destination '...' - retrying using FixedBackOff{interval=5000, currentAttempts=242, maxAttempts=unlimited}. Cause: The JMS connection has failed: ...cluster.local
2022-09-06T05:46:35.559+00:00 {taskExecutor-151} [DefaultMessageListenerContainer] [ERROR] Could not refresh JMS Connection for destination '...' - retrying using FixedBackOff{interval=5000, currentAttempts=243, maxAttempts=unlimited}. Cause: The JMS connection has failed: ...cluster.local
...
```
###### HeapDump
- I see only one `CachingConnectionFactory`
  - I see `SingleConnectionFactory$AggregatedExceptionListener` in `aggregatedExceptionListener`
  - I see `ActiveMQConnection` under `connection` field
    - I see that `exceptionListener` (on `ActiveMQConnection`) is `null` (ExceptionListener is missing - but this exception listener must be present to invoke reset of underlying JMS Connection stored on `SingleConnectionFactory`)
    - I see that `transportFailed` (on `ActiveMQConnection`) is `true` (it is a probably a reason why ExceptionListener is missing, see bellow)
    - I see that `firstFailureError` (on `ActiveMQConnection`) is `java.net.UnknownHostException` (but type of the failure is not too important)

###### Debugger
- I am able to simulate the problem in debugger
- It is enough to "delay" thread, which invokes `SingleConnectionFactory.initConnection()`
  - after new connection is created
   - but before connection is prepared, where exception listener is established

###### Code (`SingleConnectionFactory.initConnection()`)
```
public void initConnection() throws JMSException {
	if (getTargetConnectionFactory() == null) {
		throw new IllegalStateException(
				"'targetConnectionFactory' is required for lazily initializing a Connection");
	}
	synchronized (this.connectionMonitor) {
		if (this.connection != null) {
			closeConnection(this.connection);
		}
		this.connection = doCreateConnection();
		prepareConnection(this.connection);
		if (this.startedCount > 0) {
			this.connection.start();
		}
		if (logger.isDebugEnabled()) {
			logger.debug("Established shared JMS Connection: " + this.connection);
		}
	}
}
```

##### Why ExceptionListener on JMS Connection stored on `SingleConnectionFactory` is missing?
- please suppose following situation
 - the new connection is created successfully (`this.connection = doCreateConnection();`)
 - but some of subsequent methods (in my case `prepareConnection(this.connection)`) throws `JMSException`
 - we get new connection (stored on `SingleConnectionFactory`), but without JMS Exception Listener
 - this finally caused, that `SingleConnectionFactory.resetConnection()` is never invoked and "invalid" connection stays stored on `SingleConnectionFactory` (till restart or "manual" invocation of `SingleConnectionFactory.resetConnection()`)

I know it is (nearly) impossible to get this situation, because setup of exception listener (in `prepareConnection(Connection)`) should be (nearly) always done before invocation of `ActiveMQConnection.onException(IOException)` method, which sets `transportFailed` and `firstFailureError`, which finally caused that `ActiveMQConnection.setExceptionListener(ExceptionListener)` (invoked in `prepareConnection(this.connection)`) throws `ConnectionFailedException`, but I see this state in application heap dump (I am sorry).

My proposal is to change `SingleConnectionFactory.initConnection()` in a way, when new connection is created and configured/prepared in method local field and when all is ok, then assign it to instance connection field. Something like this:
```
public void initConnection() throws JMSException {
	...
	synchronized (this.connectionMonitor) {
		if (this.connection != null) {
			closeConnection(this.connection);
		}
		Connection con = doCreateConnection();
		try {
			prepareConnection(con);
			this.connection = con;
		}
		catch (JMSException ex) {
			try {
				 con.close();
			}
			catch(Throwable th) {
				logger.warn("Could not close new (but not used as shared) JMS Connection", th);
			}
			throw ex;
		}
		if (this.startedCount > 0) {
			this.connection.start();
		}
		...
	}
}
```

--- a/spring-jms/src/main/java/org/springframework/jms/connection/SingleConnectionFactory.java
+++ b/spring-jms/src/main/java/org/springframework/jms/connection/SingleConnectionFactory.java
@@ -346,8 +346,24 @@ public class SingleConnectionFactory implements ConnectionFactory, QueueConnecti
 			if (this.connection != null) {
 				closeConnection(this.connection);
 			}
-			this.connection = doCreateConnection();
-			prepareConnection(this.connection);
+			// Create new (method local) connection, which is later assigned to instance connection
+			//  - prevention to hold instance connection without exception listener, in case when
+			//    some subsequent methods (after creation of connection) throws JMSException
+			Connection con = doCreateConnection();
+			try {
+				prepareConnection(con);
+				this.connection = con;
+			}
+			catch (JMSException ex) {
+				// Attempt to close new (not used) connection to release possible resources
+				try {
+					con.close();
+				}
+				catch(Throwable th) {
+					logger.warn("Could not close new (not used as shared) JMS Connection", th);
+				}
+				throw ex;
+			}
 			if (this.startedCount > 0) {
 				this.connection.start();
 			}

--- a/spring-jms/src/main/java/org/springframework/jms/connection/SingleConnectionFactory.java
+++ b/spring-jms/src/main/java/org/springframework/jms/connection/SingleConnectionFactory.java
@@ -360,7 +360,9 @@ public class SingleConnectionFactory implements ConnectionFactory, QueueConnecti
 					con.close();
 				}
 				catch(Throwable th) {
-					logger.warn("Could not close new (not used as shared) JMS Connection", th);
+					if (logger.isDebugEnabled()) {
+						logger.debug("Could not close newly obtained JMS Connection that failed to prepare", th);
+					}
 				}
 				throw ex;
 			}

--- a/spring-jms/src/test/java/org/springframework/jms/connection/SingleConnectionFactoryTests.java
+++ b/spring-jms/src/test/java/org/springframework/jms/connection/SingleConnectionFactoryTests.java
@@ -16,6 +16,9 @@
 
 package org.springframework.jms.connection;
 
+import java.lang.reflect.Field;
+import java.util.concurrent.atomic.AtomicInteger;
+
 import jakarta.jms.Connection;
 import jakarta.jms.ConnectionFactory;
 import jakarta.jms.ExceptionListener;
@@ -29,7 +32,10 @@ import jakarta.jms.TopicConnectionFactory;
 import jakarta.jms.TopicSession;
 import org.junit.jupiter.api.Test;
 
+import org.springframework.util.ReflectionUtils;
+
 import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatExceptionOfType;
 import static org.mockito.BDDMockito.given;
 import static org.mockito.Mockito.mock;
 import static org.mockito.Mockito.times;
@@ -339,6 +345,76 @@ public class SingleConnectionFactoryTests {
 		assertThat(con.getStartCount()).isEqualTo(2);
 		assertThat(con.getCloseCount()).isEqualTo(2);
 		assertThat(listener.getCount()).isEqualTo(1);
+	}
+
+	@Test
+	public void testWithConnectionFactoryAndExceptionListenerAndReconnectOnExceptionWithJMSException() throws Exception {
+		// Throws JMSException on setExceptionListener() method, but only at the first time
+		class FailingTestConnection extends TestConnection {
+			private int setExceptionListenerInvocationCounter;
+
+			@Override
+			public void setExceptionListener(ExceptionListener exceptionListener) throws JMSException {
+				setExceptionListenerInvocationCounter++;
+				// Throw JMSException on first invocation
+				if (setExceptionListenerInvocationCounter == 1) {
+					throw new JMSException("Test JMSException (setExceptionListener())");
+				}
+				super.setExceptionListener(exceptionListener);
+			}
+		}
+
+		// Prepare base JMS ConnectionFactory
+		// - createConnection(1st) -> TestConnection,
+		// - createConnection(2nd and next) -> FailingTestConnection
+		TestConnection testCon = new TestConnection();
+		FailingTestConnection failingCon = new FailingTestConnection();
+		AtomicInteger createConnectionMethodCounter = new AtomicInteger();
+		ConnectionFactory cf = mock(ConnectionFactory.class);
+		given(cf.createConnection()).willAnswer(invocation -> {
+			int methodInvocationCounter = createConnectionMethodCounter.incrementAndGet();
+			return methodInvocationCounter == 1 ? testCon : failingCon;
+		});
+
+		// Prepare SingleConnectionFactory (setReconnectOnException())
+		// - internal connection exception listener should be registered
+		SingleConnectionFactory scf = new SingleConnectionFactory(cf);
+		scf.setReconnectOnException(true);
+		Field conField = ReflectionUtils.findField(SingleConnectionFactory.class, "connection");
+		conField.setAccessible(true);
+
+		// Get connection (1st)
+		Connection con1 = scf.getConnection();
+		assertThat(createConnectionMethodCounter.get()).isEqualTo(1);
+		assertThat(con1).isNotNull();
+		assertThat(con1.getExceptionListener()).isNotNull();
+		assertThat(con1).isSameAs(testCon);
+		// Get connection again, the same should be returned (shared connection till some problem)
+		Connection con2 = scf.getConnection();
+		assertThat(createConnectionMethodCounter.get()).isEqualTo(1);
+		assertThat(con2.getExceptionListener()).isNotNull();
+		assertThat(con2).isSameAs(con1);
+
+		// Invoke reset connection to simulate problem with connection
+		// - SCF exception listener should be invoked -> connection should be set to null
+		// - next attempt to invoke getConnection() must create new connection
+		scf.resetConnection();
+		assertThat(conField.get(scf)).isNull();
+
+		// Attempt to get connection again
+		// - JMSException should be returned from FailingTestConnection
+		// - connection should be still null (no new connection without exception listener like before fix)
+		assertThatExceptionOfType(JMSException.class).isThrownBy(() -> scf.getConnection());
+		assertThat(createConnectionMethodCounter.get()).isEqualTo(2);
+		assertThat(conField.get(scf)).isNull();
+
+		// Attempt to get connection again -> FailingTestConnection should be returned
+		//  - no JMSException is thrown, exception listener should be present
+		Connection con3 = scf.getConnection();
+		assertThat(createConnectionMethodCounter.get()).isEqualTo(3);
+		assertThat(con3).isNotNull();
+		assertThat(con3).isSameAs(failingCon);
+		assertThat(con3.getExceptionListener()).isNotNull();
 	}
 
 	@Test

